### PR TITLE
fix(nextcloud): Fix WebDAV path construction

### PR DESCRIPTION
**Pull request by provokateurin**
Fixes https://github.com/nextcloud/neon/issues/1045

### CODE REVIEW

File: packages/nextcloud/lib/src/webdav/file.dart [hunk 1]
```diff
@@ -79,17 +79,11 @@ class WebDavFile {
...
-    return segments.lastOrNull ?? '';
-  }();
+  late final String name = path.pathSegments.where((final s) => s != '').lastOrNull ?? '';
 
   /// Whether the file is hidden.
   late final bool isHidden = name.startsWith('.');
 
   /// Whether the file is a directory
-  late final bool isDirectory = (isCollection ?? false) || path.endsWith('/');
+  late final bool isDirectory = (isCollection ?? false) || path.pathSegments.last == '';
```
Comment (Leptopoda) [inline packages/nextcloud/lib/src/webdav/file.dart:88]:
```suggestion
  late final bool isDirectory = (isCollection ?? false) || path.pathSegments.last.isEmpty;
```

File: packages/nextcloud/lib/src/webdav/file.dart [hunk 2]
```diff
@@ -79,17 +79,11 @@ class WebDavFile {
...
 
   /// Returns the decoded name of the file / folder without the whole path
-  late final String name = () {
-    // normalized path (remove trailing slash)
-    final end = path.endsWith('/') ? path.length - 1 : path.length;
-    final segments = Uri.parse(path, 0, end).pathSegments;
-
-    return segments.lastOrNull ?? '';
-  }();
+  late final String name = path.pathSegments.where((final s) => s != '').lastOrNull ?? '';
```
Comment (Leptopoda) [inline packages/nextcloud/lib/src/webdav/file.dart:82]:
```suggestion
  late final String name = path.pathSegments.where((final s) => s.isNotEmpty).lastOrNull ?? '';
```

File: packages/nextcloud/lib/src/webdav/client.dart [hunk 3]
```diff
@@ -65,20 +65,11 @@ class WebDavClient {
...
-      pathBuffer
-        ..write('/')
-        ..write(path.path);
+    final segments = baseURL.pathSegments.toList()..addAll(webdavBase.pathSegments);
+    if (path != null) {
+      segments.addAll(path.pathSegments);
     }
-
-    return baseURL.replace(path: pathBuffer.toString());
+    return baseURL.replace(pathSegments: segments.where((final s) => s != ''));
```
Comment (Leptopoda) [inline packages/nextcloud/lib/src/webdav/client.dart:72]:
```suggestion
    return baseURL.replace(pathSegments: segments.where((final s) => s.isNotEmpty));
```

File: packages/nextcloud/test/webdav_test.dart [hunk 4]
```diff
@@ -9,19 +9,24 @@ import 'package:universal_io/io.dart';
...
+
+    test('constructUri $baseURL', () {
+      expect(WebDavClient.constructUri(baseURL).toString(), '$sanitizedBaseURL$webdavBase');
+      expect(WebDavClient.constructUri(baseURL, Uri(path: '/')).toString(), '$sanitizedBaseURL$webdavBase');
+      expect(WebDavClient.constructUri(baseURL, Uri(path: 'test')).toString(), '$sanitizedBaseURL$webdavBase/test');
+      expect(WebDavClient.constructUri(baseURL, Uri(path: 'test/')).toString(), '$sanitizedBaseURL$webdavBase/test');
+      expect(WebDavClient.constructUri(baseURL, Uri(path: '/test')).toString(), '$sanitizedBaseURL$webdavBase/test');
+      expect(WebDavClient.constructUri(baseURL, Uri(path: '/test/')).toString(), '$sanitizedBaseURL$webdavBase/test');
+    });
+  }
```
Comment (Leptopoda) [inline packages/nextcloud/test/webdav_test.dart:29]: Please wrap this in a group

**Review by Leptopoda — CHANGES_REQUESTED** [review]

**Comment by Leptopoda** [comment]
Without knowing how the `PathUri` is going to look I expect it to address the silly `isEmpty` stuff.
The `pathSegments` getter should by default only return relevant segments (that are not empty) and I'd like helper methods like `isDirectory`.

I already think you did so but just wanted to make it clear from the start.

**Comment by provokateurin** [comment]
Yes I added some convenience features in the PathUri to remove duplicate and ugly code.

**Review by Leptopoda — APPROVED** [review]
LGTM
Can't wait for the coming changes

**Comment by provokateurin** [comment]
Let's wait for the testing by @vauvenal5 to confirm this fixes the problems.

**Comment by vauvenal5** [comment]
Did a quick test, LGTM. :+1:

**Pull request merged**